common/persistence/client/quotas.go

```diff
@@ -76,35 +85,75 @@ var (
 func NewPriorityRateLimiter(
 	namespaceMaxQPS PersistenceNamespaceMaxQps,
 	hostMaxQPS PersistenceMaxQps,
+	perShardNamespaceMaxQPS PersistencePerShardNamespaceMaxQPS,
 	requestPriorityFn quotas.RequestPriorityFn,
 ) quotas.RequestRateLimiter {
 	hostRequestRateLimiter := newPriorityRateLimiter(
 		func() float64 { return float64(hostMaxQPS()) },
 		requestPriorityFn,
 	)
 
-	return quotas.NewNamespaceRateLimiter(func(req quotas.Request) quotas.RequestRateLimiter {
-		if req.Caller != "" && req.Caller != headers.CallerNameSystem {
-			return quotas.NewMultiRequestRateLimiter(
-				newPriorityRateLimiter(
-					func() float64 {
-						if namespaceMaxQPS == nil {
-							return float64(hostMaxQPS())
-						}
-
-						namespaceQPS := float64(namespaceMaxQPS(req.Caller))
-						if namespaceQPS <= 0 {
-							return float64(hostMaxQPS())
-						}
-
-						return namespaceQPS
-					},
-					requestPriorityFn,
-				),
-				hostRequestRateLimiter,
+	return quotas.NewMultiRequestRateLimiter(
+		newPerShardPerNamespacePriorityRateLimiter(perShardNamespaceMaxQPS, hostMaxQPS, requestPriorityFn, hostRequestRateLimiter),
+		newPriorityNamespaceRateLimiter(namespaceMaxQPS, hostMaxQPS, requestPriorityFn, hostRequestRateLimiter),
+		hostRequestRateLimiter,
+	)
+}
+
+func newPerShardPerNamespacePriorityRateLimiter(
+	perShardNamespaceMaxQPS PersistencePerShardNamespaceMaxQPS,
+	hostMaxQPS PersistenceMaxQps,
+	requestPriorityFn quotas.RequestPriorityFn,
+	hostRequestRateLimiter quotas.RequestRateLimiter,
+) quotas.RequestRateLimiter {
+	return quotas.NewMapRequestRateLimiter(func(req quotas.Request) quotas.RequestRateLimiter {
+		if hasCaller(req) && hasCallerSegment(req) {
+			return newPriorityRateLimiter(func() float64 {
+				if perShardNamespaceMaxQPS == nil || perShardNamespaceMaxQPS(req.CallerSegment) <= 0 {
+					return float64(hostMaxQPS())
+				}
+				return float64(perShardNamespaceMaxQPS(req.CallerSegment))
+			},
+				requestPriorityFn,
 			)
 		}
+		return hostRequestRateLimiter
```

Comment:
if request has no caller or caller segment, we should return no-op rate limiter? Otherwise looks like one request will go through hostRequestRateLimiter multiple times. Same for `newPriorityNamespaceRateLimiter`